### PR TITLE
Upgrade bitcoin dependency

### PR DESCRIPTION
**Pull request by tcharding**
Upgrade `rust-bitcoin` to version 0.31.0 and also other required dependencies from our stack:

- secp256k1: to v0.28.0
- internals: to v0.2.0
- bitcoind: to v0.34.0

### CODE REVIEW

**Comment by tcharding** [comment]
Includes removal of some "lol, sorta" comments, these did not make me laugh out loud and the lines marked are 100% technically correct for the function name.

File: src/miniscript/context.rs [hunk 1]
```diff
@@ -614,7 +614,7 @@ impl ScriptContext for Tap {
         // When the transaction sizes get close to block limits,
         // some guarantees are not easy to satisfy because of knapsack
         // constraints
-        if ms.ext.pk_cost > MAX_BLOCK_WEIGHT as usize {
+        if ms.ext.pk_cost as u64 > Weight::MAX_BLOCK.to_wu() {
```
Comment (apoelstra) [inline src/miniscript/context.rs:617]: In e7b5db2cc27f1c6b79abf23cffd29d13469c9245:

In a followup we should change the units of `pk_cost` to be `Weight` so we don't need to cast.

**Comment by apoelstra** [comment]
Lots of renames. ACK from me. I hope that it was possible to do this in more than 1 commit (by fixing all the actual errors, then going through one class of deprecation message at a time). But I won't ask you to redo the PR :).

File: src/miniscript/context.rs (hunk 1 above)
Comment (tcharding) [reply src/miniscript/context.rs:617]: I added it to my todo list.

**Comment by tcharding** [comment]
>  I hope that it was possible to do this in more than 1 commit (by fixing all the actual errors, then going through one class of deprecation message at a time). But I won't ask you to redo the PR :).

I can definitely do better in future, I have been doing upgrades in a very inefficient, embarrassingly manual manner. Evertime I start I think to myself "its only a few changes ..."

**Review by apoelstra — APPROVED** [review]
ACK e7b5db2cc27f1c6b79abf23cffd29d13469c9245

**Comment by apoelstra** [comment]
cc @sanket1729 can I go ahead and merge this?

**Comment by apoelstra** [comment]
I'm gonna do it and cut a release, since I think you're busy/unavailable for a bit.

**Comment by sanket1729** [comment]
@apoelstra, I am busy with some personal events for the next two weeks.

**Comment by apoelstra** [comment]
Thanks!

**Pull request merged**